crates/toolchain/tests/src/lib.rs

```diff
@@ -1,10 +1,80 @@
-//! Unit tests for OpenVM toolchain starting from rust
+use std::{
+    fs::read,
+    path::{Path, PathBuf},
+};
 
-pub mod utils;
+use eyre::Result;
+use openvm_build::{build_guest_package, get_package, is_debug, GuestOptions};
+use openvm_transpiler::{elf::Elf, openvm_platform::memory::MEM_SIZE};
+use tempfile::tempdir;
 
-#[cfg(test)]
-pub mod basic_tests;
-#[cfg(test)]
-pub mod ecc_tests;
-#[cfg(test)]
-pub mod pairing_tests;
+#[macro_export]
+macro_rules! get_programs_dir {
+    () => {{
+        let mut dir = ::std::path::PathBuf::from(env!("CARGO_MANIFEST_DIR")).to_path_buf();
+        dir.push("programs");
+        dir
+    }};
+    ($subdir:expr) => {{
+        let mut dir = ::std::path::PathBuf::from(env!("CARGO_MANIFEST_DIR")).to_path_buf();
+        dir.push($subdir);
+        dir
+    }};
+}
+
+pub fn decode_elf(elf_path: impl AsRef<Path>) -> Result<Elf> {
+    let dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+    let data = read(dir.join(elf_path))?;
+    Elf::decode(&data, MEM_SIZE as u32)
+}
+
+pub fn build_example_program(example_name: &str) -> Result<Elf> {
+    build_example_program_with_features::<&str>(example_name, [])
+}
+
+pub fn build_example_program_with_features<S: AsRef<str>>(
+    example_name: &str,
+    features: impl IntoIterator<Item = S>,
+) -> Result<Elf> {
+    let manifest_dir = get_programs_dir!();
+    build_example_program_at_path_with_features(manifest_dir, example_name, features)
+}
+
+pub fn build_example_program_at_path(manifest_dir: PathBuf, example_name: &str) -> Result<Elf> {
+    build_example_program_at_path_with_features::<&str>(manifest_dir, example_name, [])
+}
+
+pub fn build_example_program_at_path_with_features<S: AsRef<str>>(
+    manifest_dir: PathBuf,
+    example_name: &str,
+    features: impl IntoIterator<Item = S>,
+) -> Result<Elf> {
+    let pkg = get_package(manifest_dir);
+    let target_dir = tempdir()?;
+    // Build guest with default features
+    let guest_opts = GuestOptions::default()
```

Comment:
I think now some combination of https://github.com/openvm-org/openvm/blob/3ca85e83a2ab9e6a5fadfcb6f5047e171a817981/crates/toolchain/build/src/lib.rs#L352 and https://github.com/openvm-org/openvm/blob/3ca85e83a2ab9e6a5fadfcb6f5047e171a817981/crates/toolchain/build/src/lib.rs#L352 should be able to do this with less code duplication